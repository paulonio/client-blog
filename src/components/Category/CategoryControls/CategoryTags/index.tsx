import React, { FC, MouseEvent } from 'react';
import { useTranslation } from 'next-i18next';

import { CATEGORY_TAGS } from '@/constants/constants';

import styles from './styled.module.scss';
import { stringToKey } from '@/utils/utils';

interface CategorTagsProps {
  active: string[];
  onClick: (e: MouseEvent<HTMLElement>) => void;
}

const CategoryTags: FC<CategorTagsProps> = ({ active, onClick }) => {
  const { t } = useTranslation('category');
  return (
    <div className={styles.tags__wrapper}>
      {CATEGORY_TAGS.map((tag) => (
        <button
          key={tag}
          className={`${styles.tag__wrapper} ${active.includes(tag) ? styles.active : ''}`}
          onClick={onClick}
          type="button"
          data-tag={tag}
        >
          {t(stringToKey(tag))}
        </button>
      ))}
    </div>
  );
};

export default CategoryTags;
