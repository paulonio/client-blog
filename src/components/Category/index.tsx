import React, { FC } from 'react';
import { useTranslation } from 'next-i18next';

import { sen } from '@/styles/fonts';

import styles from './styled.module.scss';
import { stringToKey } from '@/utils/utils';

interface CategoryProps {
  category: string;
}

const Category: FC<CategoryProps> = ({ category }) => {
  const { t } = useTranslation(['blog', 'common']);

  return (
    <section className={styles.wrapper}>
      <div className={styles.heading}>
        <h1 className={`${sen.className} display`} data-testid="category">
          {t(stringToKey(category), { ns: 'common' })}
        </h1>
        <p className="body-s">{t('lorem-ipsum')}</p>
        <p className={`${styles.category} body-l`}>
          {t('blog')} {'>'} {t(stringToKey(category), { ns: 'common' })}
        </p>
      </div>
    </section>
  );
};

export default Category;
