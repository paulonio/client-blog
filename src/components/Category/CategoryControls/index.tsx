'use client';

import React, { ChangeEvent, FC, MouseEvent, useMemo, useState } from 'react';
import { CATEGORY_ITEMS } from '@/constants/constants';

import Categories from '@/components/CategoryBlock/Categories';
import CategoryTags from './CategoryTags';

import { sen } from '@/styles/fonts';
import { POSTS } from '@/constants/posts';

import styles from './styled.module.scss';
import Posts from '@/components/BlogPosts/Posts';
import { deleteTag, getFilteredPosts, getResults } from '@/utils/utils';

interface CategoryControlsProps {
  currentCategory: string;
}

const CategoryControls: FC<CategoryControlsProps> = ({ currentCategory }) => {
  const [activeTags, setActiveTags] = useState<string[]>([]);
  const [searchValue, setSearchValue] = useState<string>('');

  const filteredTags = useMemo(() => getResults(searchValue), [searchValue]);
  const filteredPosts = useMemo(() => getFilteredPosts(POSTS, activeTags), [activeTags]);

  const handleTagClick = (e: MouseEvent<HTMLElement>) => {
    const tag = e.currentTarget.getAttribute('data-tag');
    if (tag && activeTags.includes(tag)) {
      setActiveTags(deleteTag(activeTags, tag));
    } else if (tag) {
      setActiveTags([...activeTags, tag]);
    }
    setSearchValue('');
  };

  const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setSearchValue(value);
  };

  return (
    <section className={styles.wrapper}>
      <div className={styles.posts__wrapper}>
        <Posts posts={filteredPosts} />
      </div>
      <div>
        <div className={styles.search__bar}>
          <input
            className={`${styles.input} ${sen.className}`}
            type="text"
            placeholder="Search for tag..."
            value={searchValue}
            onChange={handleSearchChange}
          />
          <button className={`${styles.button} ${sen.className}`} type="button">
            Search
          </button>
          {searchValue && (
            <div className={styles.results}>
              {filteredTags.length > 0 ? (
                filteredTags.map((tag) => (
                  <div
                    key={tag}
                    className={styles.results__item}
                    data-tag={tag}
                    onClick={handleTagClick}
                    aria-hidden
                  >
                    {tag}
                  </div>
                ))
              ) : (
                <div>No result</div>
              )}
            </div>
          )}
        </div>
        <div className={styles.categories__wrapper}>
          <h2 className={sen.className}>Categories</h2>
          <Categories
            type="controls"
            currentCategory={currentCategory}
            categories={CATEGORY_ITEMS}
          />
        </div>
        <div className={styles.tags__wrapper}>
          <h2 className={sen.className}>All Tags</h2>
          <CategoryTags active={activeTags} onClick={handleTagClick} />
        </div>
      </div>
    </section>
  );
};

export default CategoryControls;
