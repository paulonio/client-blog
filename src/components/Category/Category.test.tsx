import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Category from '.';
import CategoryControls from './CategoryControls';

const TEST_VALUE = 'test';

describe('Category', () => {
  it('should render category', () => {
    const { container } = render(<Category category={TEST_VALUE} />);

    expect(container).toMatchSnapshot();
  });

  it('should display correct category', () => {
    render(<Category category={TEST_VALUE} />);

    const title = screen.getByTestId('category');

    expect(title).toHaveTextContent('test');
  });

  it('should render controls correctly', () => {
    const { container } = render(<CategoryControls currentCategory={TEST_VALUE} />);

    expect(container).toMatchSnapshot();
  });

  it('should handle change input', () => {
    render(<CategoryControls currentCategory={TEST_VALUE} />);

    const input: HTMLInputElement = screen.getByTestId('search-input');
    fireEvent.change(input, { target: { value: TEST_VALUE } });

    expect(input.value).toEqual(TEST_VALUE);
  });

  it('should handle change input', () => {
    render(<CategoryControls currentCategory={TEST_VALUE} />);

    const result = screen.queryByTestId('result');
    expect(result).toBe(null);
    const input: HTMLInputElement = screen.getByTestId('search-input');
    fireEvent.change(input, { target: { value: TEST_VALUE } });
    const newResult = screen.getByTestId('result');
    expect(newResult).toBeInTheDocument();
    fireEvent.change(input, { target: { value: 'business' } });
    const trueResult = screen.getByTestId('true-result');
    expect(trueResult).toHaveTextContent(/business/gi);
  });
});
