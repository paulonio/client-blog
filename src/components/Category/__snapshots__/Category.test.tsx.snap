// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category should render category 1`] = `
<div>
  <section
    class="wrapper"
  >
    <div
      class="heading"
    >
      <h1
        class="className display"
        data-testid="category"
      >
        test
      </h1>
      <p
        class="body-s"
      >
        lorem-ipsum
      </p>
      <p
        class="category body-l"
      >
        blog
         
        &gt;
         
        test
      </p>
    </div>
  </section>
</div>
`;

exports[`Category should render controls correctly 1`] = `
<div>
  <section
    class="wrapper"
  >
    <div
      class="posts__wrapper"
    >
      <div
        class="posts__wrapper"
      >
        <h2
          class="message"
        >
          no-result
        </h2>
      </div>
    </div>
    <div>
      <div
        class="search__bar"
      >
        <input
          class="input className"
          data-testid="search-input"
          placeholder="search-tag"
          type="text"
          value=""
        />
        <button
          class="button className"
          type="button"
        >
          search
        </button>
      </div>
      <div
        class="categories__wrapper"
      >
        <h2
          class="className"
        >
          categories
        </h2>
        <div
          class="categories__wrapper controls"
        >
          <a
            class="link"
            href="/business"
          >
            <div
              class="category__wrapper   controls"
            >
              <div
                class="icon__wrapper controls"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M13.0305 1.91382L2.04 0.244322C1.53 0.164822 1.0095 0.304322 0.6195 0.634322C0.2295 0.974822 0 1.46382 0 1.97382V23.0443C0 23.5948 0.45 24.0448 1.0005 24.0448H4.251V18.7948C4.251 17.8243 5.031 17.0443 6.0015 17.0443H8.502C9.4725 17.0443 10.2525 17.8243 10.2525 18.7948V24.0448H14.502V3.64482C14.5005 2.78382 13.8795 2.05482 13.0305 1.91382V1.91382ZM5.5005 14.7943H4.0005C3.5865 14.7943 3.2505 14.4583 3.2505 14.0443C3.2505 13.6303 3.5865 13.2943 4.0005 13.2943H5.5005C5.9145 13.2943 6.2505 13.6303 6.2505 14.0443C6.2505 14.4583 5.9145 14.7943 5.5005 14.7943ZM5.5005 11.7943H4.0005C3.5865 11.7943 3.2505 11.4583 3.2505 11.0443C3.2505 10.6303 3.5865 10.2943 4.0005 10.2943H5.5005C5.9145 10.2943 6.2505 10.6303 6.2505 11.0443C6.2505 11.4583 5.9145 11.7943 5.5005 11.7943ZM5.5005 8.79432H4.0005C3.5865 8.79432 3.2505 8.45832 3.2505 8.04432C3.2505 7.63032 3.5865 7.29432 4.0005 7.29432H5.5005C5.9145 7.29432 6.2505 7.63032 6.2505 8.04432C6.2505 8.45832 5.9145 8.79432 5.5005 8.79432ZM5.5005 5.79432H4.0005C3.5865 5.79432 3.2505 5.45832 3.2505 5.04432C3.2505 4.63032 3.5865 4.29432 4.0005 4.29432H5.5005C5.9145 4.29432 6.2505 4.63032 6.2505 5.04432C6.2505 5.45832 5.9145 5.79432 5.5005 5.79432ZM10.5 14.7943H9C8.586 14.7943 8.25 14.4583 8.25 14.0443C8.25 13.6303 8.586 13.2943 9 13.2943H10.5C10.914 13.2943 11.25 13.6303 11.25 14.0443C11.25 14.4583 10.914 14.7943 10.5 14.7943ZM10.5 11.7943H9C8.586 11.7943 8.25 11.4583 8.25 11.0443C8.25 10.6303 8.586 10.2943 9 10.2943H10.5C10.914 10.2943 11.25 10.6303 11.25 11.0443C11.25 11.4583 10.914 11.7943 10.5 11.7943ZM10.5 8.79432H9C8.586 8.79432 8.25 8.45832 8.25 8.04432C8.25 7.63032 8.586 7.29432 9 7.29432H10.5C10.914 7.29432 11.25 7.63032 11.25 8.04432C11.25 8.45832 10.914 8.79432 10.5 8.79432ZM10.5 5.79432H9C8.586 5.79432 8.25 5.45832 8.25 5.04432C8.25 4.63032 8.586 4.29432 9 4.29432H10.5C10.914 4.29432 11.25 4.63032 11.25 5.04432C11.25 5.45832 10.914 5.79432 10.5 5.79432Z"
                    fill="#232536"
                  />
                  <path
                    d="M22.6205 10.887L15.5 9.396V24.045H22.25C23.2145 24.045 24.0005 23.2605 24.0005 22.2945V12.597C24.0005 11.7705 23.4365 11.067 22.6205 10.887ZM20.2505 21.045H18.7505C18.3365 21.045 18.0005 20.709 18.0005 20.295C18.0005 19.881 18.3365 19.545 18.7505 19.545H20.2505C20.6645 19.545 21.0005 19.881 21.0005 20.295C21.0005 20.709 20.6645 21.045 20.2505 21.045ZM20.2505 18.045H18.7505C18.3365 18.045 18.0005 17.709 18.0005 17.295C18.0005 16.881 18.3365 16.545 18.7505 16.545H20.2505C20.6645 16.545 21.0005 16.881 21.0005 17.295C21.0005 17.709 20.6645 18.045 20.2505 18.045ZM20.2505 15.045H18.7505C18.3365 15.045 18.0005 14.709 18.0005 14.295C18.0005 13.881 18.3365 13.545 18.7505 13.545H20.2505C20.6645 13.545 21.0005 13.881 21.0005 14.295C21.0005 14.709 20.6645 15.045 20.2505 15.045Z"
                    fill="#FFD050"
                  />
                </svg>
              </div>
              <h3
                class="className"
              >
                business
              </h3>
            </div>
          </a>
          <a
            class="link"
            href="/startup"
          >
            <div
              class="category__wrapper   controls"
            >
              <div
                class="icon__wrapper controls"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M5.85054 6.21152L3.85792 6.36546C3.30345 6.40832 2.80963 6.69892 2.50299 7.16273L0.204943 10.6384C-0.0276155 10.9901 -0.0645859 11.4303 0.105981 11.8159C0.276593 12.2015 0.627205 12.4703 1.04385 12.5348L2.86834 12.8174C3.29514 10.5524 4.31683 8.29305 5.85054 6.21152Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M10.1836 20.1316L10.4662 21.9561C10.5307 22.3728 10.7995 22.7234 11.185 22.8939C11.3454 22.9649 11.5152 22.9999 11.6842 22.9999C11.9215 22.9999 12.1572 22.9308 12.3626 22.795L15.8384 20.497C16.3022 20.1903 16.5928 19.6964 16.6355 19.142L16.7895 17.1494C14.7079 18.6832 12.4486 19.7049 10.1836 20.1316Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M9.49658 18.8667C9.55992 18.8667 9.62362 18.8614 9.68718 18.8508C10.6361 18.6922 11.5507 18.4249 12.4232 18.075L4.92442 10.5763C4.57457 11.4487 4.30729 12.3633 4.14863 13.3123C4.0861 13.6863 4.21156 14.0674 4.4797 14.3356L8.66387 18.5198C8.8865 18.7424 9.18698 18.8667 9.49658 18.8667Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M21.1661 10.1973C23.0011 6.64969 23.069 2.90072 22.9731 1.14249C22.9403 0.540047 22.46 0.0597456 21.8575 0.026908C21.5712 0.0112753 21.2319 0 20.8495 0C18.8835 0 15.7726 0.297785 12.8027 1.83397C10.4424 3.0548 7.34745 5.74295 5.52148 9.27567C5.54305 9.29251 5.56411 9.31039 5.58397 9.33025L13.6698 17.4161C13.6897 17.436 13.7075 17.457 13.7244 17.4786C17.2571 15.6525 19.9453 12.5576 21.1661 10.1973ZM13.3731 4.86231C14.6866 3.54876 16.8241 3.54862 18.1377 4.86231C18.7741 5.49862 19.1245 6.34472 19.1245 7.24463C19.1245 8.14454 18.7741 8.99064 18.1377 9.62695C17.481 10.2837 16.618 10.6121 15.7554 10.6122C14.8925 10.6122 14.03 10.2838 13.3731 9.62695C12.7367 8.99064 12.3863 8.14454 12.3863 7.24463C12.3863 6.34472 12.7367 5.49862 13.3731 4.86231Z"
                    fill="#232536"
                  />
                  <path
                    d="M14.3264 8.67403C15.1146 9.46218 16.3971 9.46223 17.1853 8.67403C17.5671 8.29219 17.7773 7.78458 17.7773 7.24462C17.7773 6.70466 17.5671 6.19704 17.1853 5.81525C16.7912 5.42115 16.2735 5.22412 15.7559 5.22412C15.2382 5.22412 14.7205 5.42115 14.3265 5.81525C13.9447 6.19704 13.7344 6.70466 13.7344 7.24462C13.7344 7.78458 13.9447 8.29224 14.3264 8.67403Z"
                    fill="#232536"
                  />
                  <path
                    d="M0.687523 18.954C0.859978 18.954 1.03243 18.8882 1.16396 18.7566L3.36392 16.5566C3.62708 16.2935 3.62708 15.8668 3.36392 15.6037C3.10082 15.3405 2.67415 15.3405 2.411 15.6037L0.211036 17.8036C-0.0521162 18.0668 -0.0521162 18.4934 0.211036 18.7566C0.342612 18.8881 0.515068 18.954 0.687523 18.954Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M5.37991 17.6201C5.1168 17.357 4.69014 17.357 4.42698 17.6201L0.197364 21.8497C-0.0657881 22.1129 -0.0657881 22.5395 0.197364 22.8027C0.32894 22.9342 0.501396 23 0.673851 23C0.846306 23 1.01876 22.9342 1.15029 22.8026L5.37987 18.5731C5.64306 18.3099 5.64306 17.8833 5.37991 17.6201Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M6.4422 19.6358L4.24229 21.8358C3.97913 22.099 3.97913 22.5256 4.24229 22.7887C4.37386 22.9203 4.54632 22.9861 4.71873 22.9861C4.89114 22.9861 5.06364 22.9203 5.19517 22.7887L7.39513 20.5888C7.65828 20.3256 7.65828 19.899 7.39513 19.6358C7.13202 19.3727 6.70535 19.3727 6.4422 19.6358Z"
                    fill="#FFD050"
                  />
                </svg>
              </div>
              <h3
                class="className"
              >
                startup
              </h3>
            </div>
          </a>
          <a
            class="link"
            href="/economy"
          >
            <div
              class="category__wrapper   controls"
            >
              <div
                class="icon__wrapper controls"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M5.11557 15.4253H2.53171C1.5999 15.4253 0.841797 16.2179 0.841797 17.192V22.915C0.841797 23.5133 1.30738 24 1.87959 24H5.76759C6.33985 24 6.80544 23.5133 6.80544 22.915V17.1921C6.80544 16.2179 6.04738 15.4253 5.11557 15.4253Z"
                    fill="#232536"
                  />
                  <path
                    d="M12.6996 11.9491H10.1157C9.18389 11.9491 8.42578 12.7417 8.42578 13.7158V22.9147C8.42578 23.513 8.89137 23.9997 9.46363 23.9997H13.3516C13.9239 23.9997 14.3894 23.513 14.3894 22.9147V13.7158C14.3894 12.7416 13.6314 11.9491 12.6996 11.9491Z"
                    fill="#232536"
                  />
                  <path
                    d="M20.4261 7.98877H17.8423C16.9104 7.98877 16.1523 8.78133 16.1523 9.75545V22.9148C16.1523 23.5131 16.6179 23.9998 17.1901 23.9998H21.0781C21.6504 23.9998 22.1159 23.5131 22.1159 22.9148V9.75545C22.116 8.78133 21.3579 7.98877 20.4261 7.98877Z"
                    fill="#232536"
                  />
                  <path
                    d="M18.3681 0.668189C18.052 0.201126 17.538 -0.045859 16.9929 0.00645363L12.1128 0.478486C11.5678 0.53122 11.1059 0.87294 10.8771 1.39264C10.6483 1.91235 10.7004 2.50152 11.0165 2.96858L11.2889 3.37105L5.23176 7.7468L5.22557 7.75134C4.79576 8.0693 4.51015 8.54325 4.42133 9.08583C4.33251 9.62841 4.45106 10.1746 4.75519 10.624C5.05932 11.0733 5.51266 11.3719 6.03165 11.4648C6.1441 11.4849 6.25669 11.4949 6.36851 11.4949C6.77133 11.4949 7.1637 11.3657 7.49957 11.1184L13.5664 6.73575L13.8476 7.15111C14.1357 7.57669 14.588 7.81983 15.0783 7.81983C15.1261 7.81983 15.1743 7.81748 15.2228 7.81284C15.7677 7.76011 16.2297 7.41834 16.4585 6.89859L18.5076 2.24413C18.7364 1.72433 18.6843 1.13521 18.3681 0.668189Z"
                    fill="#FFD050"
                  />
                </svg>
              </div>
              <h3
                class="className"
              >
                economy
              </h3>
            </div>
          </a>
          <a
            class="link"
            href="/technology"
          >
            <div
              class="category__wrapper   controls"
            >
              <div
                class="icon__wrapper controls"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M15.5176 7.03125C15.5176 7.80647 16.1486 8.4375 16.9238 8.4375H21.1042C21.0283 7.44788 20.7988 6.50213 20.4316 5.625H16.9238C16.1486 5.625 15.5176 6.25603 15.5176 7.03125Z"
                    fill="#232536"
                  />
                  <path
                    d="M6.37695 24H14.8145C15.2031 24 15.5176 23.6855 15.5176 23.2969V19.6875H16.9238V15.1172L9.893 9.84375H3.56445C3.17581 9.84375 2.86133 10.1582 2.86133 10.5469C2.86133 13.0373 3.88236 15.4228 5.67383 17.1428V23.2969C5.67383 23.6855 5.98831 24 6.37695 24Z"
                    fill="#FFD050"
                  />
                  <path
                    d="M12.002 0C11.6133 0 11.2988 0.314484 11.2988 0.703125V1.40625H8.35681C8.06558 0.589594 7.29242 0 6.37695 0C5.2138 0 4.26758 0.946219 4.26758 2.10938C4.26758 3.27253 5.2138 4.21875 6.37695 4.21875C7.29242 4.21875 8.06558 3.62916 8.35681 2.8125H11.2988V5.625H4.13806C3.84683 4.80834 3.07367 4.21875 2.1582 4.21875C0.995047 4.21875 0.00195312 5.16497 0.00195312 6.32812C0.00195312 7.49128 0.995047 8.4375 2.1582 8.4375C3.07367 8.4375 3.84683 7.84791 4.13806 7.03125H11.2988V9.14095L18.0485 14.2033C18.2257 14.3358 18.3301 14.5445 18.3301 14.7656V19.6875H19.0332C20.1964 19.6875 21.1426 18.7413 21.1426 17.5781V14.0625H23.2988C23.5584 14.0625 23.7966 13.9197 23.9189 13.691C24.0411 13.4624 24.0274 13.185 23.8839 12.9693L21.7533 9.84375H16.9238C15.3727 9.84375 14.1113 8.58239 14.1113 7.03125C14.1113 5.48011 15.3727 4.21875 16.9238 4.21875H19.6897C18.0628 1.68647 15.2291 0 12.002 0V0Z"
                    fill="#232536"
                  />
                </svg>
              </div>
              <h3
                class="className"
              >
                technology
              </h3>
            </div>
          </a>
        </div>
      </div>
      <div
        class="tags__wrapper"
      >
        <h2
          class="className"
        >
          all-tags
        </h2>
        <div
          class="tags__wrapper"
        >
          <button
            class="tag__wrapper "
            data-tag="Business"
            type="button"
          >
            business
          </button>
          <button
            class="tag__wrapper "
            data-tag="Experience"
            type="button"
          >
            experience
          </button>
          <button
            class="tag__wrapper "
            data-tag="Screen"
            type="button"
          >
            screen
          </button>
          <button
            class="tag__wrapper "
            data-tag="Technology"
            type="button"
          >
            technology
          </button>
          <button
            class="tag__wrapper "
            data-tag="Marketing"
            type="button"
          >
            marketing
          </button>
          <button
            class="tag__wrapper "
            data-tag="Life"
            type="button"
          >
            life
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;
